File: Flutter/ios/Classes/tools/FTXV2LiveTools.m
// Copyright (c) 2022 Tencent. All rights reserved.

#import "FTXV2LiveTools.h"
#import "FTXEvent.h"

@implementation FTXV2LiveTools

+ (V2TXLiveRotation)transRotationFromDegree:(int)rotation {
    V2TXLiveRotation rotationCode = V2TXLiveRotation270;
    if (rotation <= 0) {
        rotationCode = V2TXLiveRotation0;
    } else if (rotation <= 90) {
        rotationCode = V2TXLiveRotation90;
    } else if (rotation <= 180) {
        rotationCode = V2TXLiveRotation180;
    } else {
        rotationCode = V2TXLiveRotation270;
    }
    return rotationCode;
}

+ (NSDictionary *)buildNetBundle:(V2TXLivePlayerStatistics *)statistics {
    NSMutableDictionary *dic = @{}.mutableCopy;
    [dic setValue:@(statistics.appCpu) forKey:NET_STATUS_CPU_USAGE];
    [dic setValue:@(statistics.width) forKey:NET_STATUS_VIDEO_WIDTH];
    [dic setValue:@(statistics.height) forKey:NET_STATUS_VIDEO_HEIGHT];
    [dic setValue:@(statistics.netSpeed) forKey:NET_STATUS_NET_SPEED];
    [dic setValue:@(statistics.fps) forKey:NET_STATUS_VIDEO_FPS];
    [dic setValue:@(statistics.videoBitrate) forKey:NET_STATUS_VIDEO_BITRATE];
    [dic setValue:@(statistics.audioBitrate) forKey:NET_STATUS_AUDIO_BITRATE];
    [dic setValue:@(statistics.jitterBufferDelay) forKey:NET_STATUS_NET_JITTER];
    [dic setValue:@(statistics.systemCpu) forKey:NET_STATUS_SYSTEM_CPU];
    [dic setValue:@(statistics.videoPacketLoss) forKey:NET_STATUS_VIDEO_LOSS];
    [dic setValue:@(statistics.audioPacketLoss) forKey:NET_STATUS_AUDIO_LOSS];
    [dic setValue:@(statistics.audioTotalBlockTime) forKey:NET_STATUS_AUDIO_TOTAL_BLOCK_TIME];
    [dic setValue:@(statistics.videoTotalBlockTime) forKey:NET_STATUS_VIDEO_TOTAL_BLOCK_TIME];
    [dic setValue:@(statistics.videoBlockRate) forKey:NET_STATUS_VIDEO_BLOCK_RATE];
    [dic setValue:@(statistics.audioBlockRate) forKey:NET_STATUS_AUDIO_BLOCK_RATE];
    [dic setValue:@(statistics.rtt) forKey:NET_STATUS_RTT];
    return dic;;
}

@end

File: Flutter/android/src/main/java/com/tencent/vod/flutter/tools/FTXV2LiveTools.java
package com.tencent.vod.flutter.tools;

import android.graphics.Bitmap;
import android.graphics.ImageFormat;
import android.graphics.Rect;
import android.graphics.YuvImage;
import android.os.Bundle;

import com.tencent.live2.V2TXLiveDef;
import com.tencent.vod.flutter.FTXEvent;

import java.io.ByteArrayOutputStream;
import java.nio.ByteBuffer;

public class FTXV2LiveTools {

    public static V2TXLiveDef.V2TXLiveRotation transRotationFromDegree(int rotation) {
        V2TXLiveDef.V2TXLiveRotation rotationCode;
        if (rotation <= 0) {
            rotationCode = V2TXLiveDef.V2TXLiveRotation.V2TXLiveRotation0;
        } else if (rotation <= 90) {
            rotationCode = V2TXLiveDef.V2TXLiveRotation.V2TXLiveRotation90;
        } else if (rotation <= 180) {
            rotationCode = V2TXLiveDef.V2TXLiveRotation.V2TXLiveRotation180;
        } else {
            rotationCode = V2TXLiveDef.V2TXLiveRotation.V2TXLiveRotation270;
        }
        return rotationCode;
    }

    public static ByteBuffer yuv420ToARGB8888(ByteBuffer yuv420Buffer, int width, int height) {
        // 将YUV420格式的ByteBuffer转换为byte数组
        byte[] yuv420Bytes = new byte[yuv420Buffer.remaining()];
        yuv420Buffer.get(yuv420Bytes);

        // 将YUV420格式的byte数组转换为Bitmap
        YuvImage yuvImage = new YuvImage(yuv420Bytes, ImageFormat.YUV_420_888, width, height, null);
        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
        Rect rect = new Rect(0, 0, width, height);
        yuvImage.compressToJpeg(rect, 100, outputStream);
        byte[] argbBytes = outputStream.toByteArray();

        // 将ARGB8888格式的byte数组转换为ByteBuffer
        return ByteBuffer.wrap(argbBytes);
    }

    public static Bundle buildNetBundle(V2TXLiveDef.V2TXLivePlayerStatistics statistics) {
        Bundle bundle = new Bundle();
        if (null != statistics) {
            bundle.putInt(FTXEvent.TUINetConst.NET_STATUS_CPU_USAGE, statistics.appCpu);
            bundle.putInt(FTXEvent.TUINetConst.NET_STATUS_VIDEO_WIDTH, statistics.width);
            bundle.putInt(FTXEvent.TUINetConst.NET_STATUS_VIDEO_HEIGHT, statistics.height);
            bundle.putInt(FTXEvent.TUINetConst.NET_STATUS_NET_SPEED, statistics.netSpeed);
            bundle.putInt(FTXEvent.TUINetConst.NET_STATUS_VIDEO_FPS, statistics.fps);
            bundle.putInt(FTXEvent.TUINetConst.NET_STATUS_VIDEO_BITRATE, statistics.videoBitrate);
            bundle.putInt(FTXEvent.TUINetConst.NET_STATUS_AUDIO_BITRATE, statistics.audioBitrate);
            bundle.putInt(FTXEvent.TUINetConst.NET_STATUS_NET_JITTER, statistics.jitterBufferDelay);
            bundle.putInt(FTXEvent.TUINetConst.NET_STATUS_SYSTEM_CPU, statistics.systemCpu);
            bundle.putInt(FTXEvent.TUINetConst.NET_STATUS_VIDEO_LOSS, statistics.videoPacketLoss);
            bundle.putInt(FTXEvent.TUINetConst.NET_STATUS_AUDIO_LOSS, statistics.audioPacketLoss);
            bundle.putInt(FTXEvent.TUINetConst.NET_STATUS_AUDIO_TOTAL_BLOCK_TIME, statistics.audioTotalBlockTime);
            bundle.putInt(FTXEvent.TUINetConst.NET_STATUS_VIDEO_TOTAL_BLOCK_TIME, statistics.videoTotalBlockTime);
            bundle.putInt(FTXEvent.TUINetConst.NET_STATUS_VIDEO_BLOCK_RATE, statistics.videoBlockRate);
            bundle.putInt(FTXEvent.TUINetConst.NET_STATUS_AUDIO_BLOCK_RATE, statistics.audioBlockRate);
            bundle.putInt(FTXEvent.TUINetConst.NET_STATUS_RTT, statistics.rtt);
        }
        return bundle;
    }

}
